### DISCUSSION TITLE
Counter input improvement

### DISCUSSION THREAD

**Discussion by vitalyga** [arendst/Tasmota#12982] [Ideas]
### PROBLEM DESCRIPTION
The current approach of the "Counter" input makes it difficult and sometimes impossible to get required information from device. 
By the "current approach" I mean that we have currently the CounterType param that could be either pulse counter or pulse timer but not the both. I don't see any logical reason why it can't be both pulse counter and pulse timer. 
For example, I have a simple power meter that sends one pulse per N W/h. So the pulse timer is useful if I want to calculate the current power consumption. But at the same time I want also to know the pulse counter value in order to calculate the Total Power Consumption since the certain time point (since the last payment for example). In theory, I can connect the same wire to two gpios and set one to CounterType 0 and second to CounterType 1 but why not get both values per same input? Also, it's not always possible to change the hardware part of the system.
So, I would suggest to drop the CounterType param and always provide both pulse counter and pulse timer values.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [ ] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [ ] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [ ] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [ ] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [ ] Device used (e.g., Sonoff Basic): _____
- [ ] Tasmota binary firmware version number used: _____
  - [ ] Pre-compiled
  - [ ] Self-compiled
- [ ] Flashing tools used: _____
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by arendst** [comment]
>  I don't see any logical reason why it can't be both pulse counter and pulse timer

The logical reason is that due to memory constraints a register can only be either pulse counter OR pulse timer.

**Reply by vitalyga** [reply]
Sorry but I'm not sure I understand you. Why just not use one ulong value for counter and another for timer? Is there 8 additional bytes is such a big deal? Instead of dropping CounterType we can have 3 options: 0 for pulse counter, 1 for pulse timer and 2 for both.

**Reply by sfromis** [reply]
You can easily have both count and time at the same time. Just configure two pins, and connect the input to both.

**Reply by vitalyga** [reply]
Did you read my initial post? Especially, this part:
"In theory, I can connect the same wire to two gpios and set one to CounterType 0 and second to CounterType 1 but why not get both values per same input? Also, it's not always possible to change the hardware part of the system."
BTW, @arendst - for two pins connected there is no "memory constraints"? Are you sure your "logical" reason is really logical? ;-)

**Reply by sfromis** [reply]
As the counters/timers are in persistent memory, allocated from a limited size block used for lots of different purposes, doubling the memory needed for the whole set of counters (what you suggest) is a significant cost, especially when you have a very feasible way of getting what you want without any extra code needed.

**Reply by vitalyga** [reply]
@sfromis You want to tell me that while having 4M of flash memory and single connected pin for the whole MCU I can't afford to have additional 8 bytes of persistent memory? Is it joke or something?
The best solution is give user as much options as possible and let him to decide how he want to use memory of his own device. If there is only one gpio pin in use, there plenty of available resources and there is no any problem to use additional couple of bytes. Especially when we are talking about almost already written functionality that just need to be a little improved.

**Reply by sfromis** [reply]
It would double the static memory needed for all counters, of which you have multiple. Much more than "a couple of bytes". Additionally, it would add a bit of code complexity. It is not a dynamic area, but a fixed structure with static offsets, designed to allow firmware updating with all data still persisting. Not up to what the user can decide, unless the user decides to change the source code significantly.

**Reply by vitalyga** [reply]
You know, literally ANY feature one way or another will add complexity and memory use. In my initial post I described a real life example and suggested the feature that gives a user the solution for this case scenario. 
Today everyone that have simple pulse-based power meter in their home forced to use ESPHome project because of lack of this functionality in tasmota. Personally, I like tasmota and would prefer to continue with the project. And I really don't understand why not to make this little improvement that will make a whole system much more flexible and usable.

**Reply by sfromis** [reply]
Why not connect the input to two pins? Tasmota has the functionality you ask for.

**Reply by vitalyga** [reply]
1. Very limited amount of users will want to mess with the devices they bought to make such customization inside of it. You know, not all people are using tasmota in their DIY projects. I think the majority are people that bought device like Shelly or Sonoff and want to make it a little bit more flexible.
2. In my case, I had the device with this functionality that I implemented by myself. In some point I decided to unify all devices I have and put tasmota in all. The problem of "why not connect the input to two pins?" is that the device is installed in the not so accessible place and I really don't want to uninstall it, then make the hardware change and then install it back. I think I'm not only person in the world with such case scenario.

**Reply by vitalyga** [reply]
3. What is second pin you can connect in ESP-01 that I'm using?

**Reply by sfromis** [reply]
I'd not use an ESP-01 for anything, but in general, you can liberate tx/rx for other connections.

Anyway, if someone is in the mood, it could be doable to create variation to halve the number of available counters, to have each double the function.

**Reply by vitalyga** [reply]
> I'd not use an ESP-01 for anything

Why not? For a single pin relay switch or one pin input ESP-01 is the perfect solution. I'm using several of such devices for years and haven't single issue with it. Until I decided to flash tasmota and found out that tasmota is not capable of counting pulses and measure time interval at the same time. Who would have thought...

**Reply by sfromis** [reply]
I've seen users of ESP-01 have too much trouble with variations, that's why. Anyway, you do have an extra pin there, as I suggested.

**Reply by vitalyga** [reply]
Yep, thanks for that, I didn't know of such possibility. May be I will try it in my future projects but not with the power meter I have right now due to the said above.
